source/opt/constant_value.h

```diff
@@ -0,0 +1,321 @@
+  const FloatConst* AsFloatConst() const override { return this; }
+
+  std::unique_ptr<FloatConst> CopyFloatConst() const {
+    return std::unique_ptr<FloatConst>(new FloatConst(type_, words_));
+  }
+  std::unique_ptr<Const> Copy() const override {
+    return std::unique_ptr<Const>(CopyFloatConst().release());
+  }
+
+  std::vector<uint32_t> words() const override { return words_; }
+
+ private:
+  std::vector<uint32_t> words_;
+};
+
+// Bool type constant.
+class BoolConst : public ScalarConst {
+ public:
+  BoolConst(const analysis::Type* ty, bool val) : ScalarConst(ty), val_(val) {}
+
+  BoolConst* AsBoolConst() override { return this; }
+  const BoolConst* AsBoolConst() const override { return this; }
+
+  std::unique_ptr<BoolConst> CopyBoolConst() const {
+    return std::unique_ptr<BoolConst>(new BoolConst(type_, val_));
+  }
+  std::unique_ptr<Const> Copy() const override {
+    return std::unique_ptr<Const>(CopyBoolConst().release());
+  }
+
+  std::vector<uint32_t> words() const override {
+    return {static_cast<uint32_t>(val_)};
+  }
+
+  bool val() const { return val_; }
+
+ private:
+  bool val_;
+};
+
+// Abstract class for composite constants.
+class CompositeConst : public Const {
+ public:
+  CompositeConst() = delete;
+  CompositeConst* AsCompositeConst() override { return this; }
+  const CompositeConst* AsCompositeConst() const override { return this; }
+
+  virtual uint32_t NumComponents() const = 0;
+  virtual const Const* GetComponent(uint32_t index) const = 0;
+  virtual void AppendComponent(const Const* c) = 0;
+
+ protected:
+  CompositeConst(const analysis::Type* ty) : Const(ty) {}
+};
+
+// Struct type constant.
+class StructConst : public CompositeConst {
+ public:
+  StructConst(const analysis::Type* ty) : CompositeConst(ty), components_() {}
+  StructConst(const analysis::Type* ty,
+              const std::vector<const Const*>& components)
+      : CompositeConst(ty), components_(components) {}
+  StructConst(const analysis::Type* ty, std::vector<const Const*>&& components)
+      : CompositeConst(ty), components_(components) {}
+
+  StructConst* AsStructConst() override { return this; }
+  const StructConst* AsStructConst() const override { return this; }
+
+  std::unique_ptr<StructConst> CopyStructConst() const {
+    return std::unique_ptr<StructConst>(new StructConst(type_, components_));
+  }
+  std::unique_ptr<Const> Copy() const override {
+    return std::unique_ptr<Const>(CopyStructConst().release());
+  }
+
+  uint32_t NumComponents() const override {
+    return static_cast<uint32_t>(components_.size());
+  }
+  const Const* GetComponent(uint32_t index) const override {
+    assert(static_cast<size_t>(index) < components_.size() &&
+           "GetComponent() index out of bound.");
+    return components_[index];
+  }
+  void AppendComponent(const Const* c) override { components_.push_back(c); }
+
+ private:
+  std::vector<const Const*> components_;
```

Comment:
It seems this can be moved to `CompositeConst` so that some of the methods in `StructConst`, `ArraryConst`, `VectorConst` can also be moved to `CompositeConst`?